models/spike_train_injector.cpp

```diff
@@ -0,0 +1,463 @@
+
+void
+spike_train_injector::init_state_()
+{
+  Node::init_state_();
+}
+
+void
+spike_train_injector::init_buffers_()
+{
+}
+
+void
+spike_train_injector::pre_run_hook()
+{
+  // We do not need to recalibrate time objects, since they are
+  // recalibrated on instance construction and resolution cannot
+  // change after a single node instance has been created.
+
+  // Off-grid communication needs to be activated here since this model
+  // is not an exclusive precise spiking model
+  if ( is_off_grid() )
+  {
+    kernel().event_delivery_manager.set_off_grid_communication( true );
+    LOG( M_INFO,
+      "spike_train_injector::pre_run_hook",
+      "Static injecor neuron has been configured to emit precisely timed "
+      "spikes: the kernel property off_grid_spiking has been set to true.\n\n"
+      "NOTE: Mixing precise-spiking and normal neuron models may "
+      "lead to inconsistent results." );
+  }
+
+  // by adding time objects, all overflows will be handled gracefully
+  V_.t_min_ = ( P_.origin_ + P_.start_ ).get_steps();
+  V_.t_max_ = ( P_.origin_ + P_.stop_ ).get_steps();
+}
+
+
+/* ----------------------------------------------------------------
+ * Other functions
+ * ---------------------------------------------------------------- */
+void
+spike_train_injector::update( Time const& sliceT0, const long from, const long to )
+{
+  if ( P_.spike_stamps_.empty() )
+  {
+    return;
+  }
+
+  assert( not P_.precise_times_ or P_.spike_stamps_.size() == P_.spike_offsets_.size() );
+  assert( P_.spike_multiplicities_.empty() or P_.spike_stamps_.size() == P_.spike_multiplicities_.size() );
+
+  const Time tstart = sliceT0 + Time::step( from );
+  const Time tstop = sliceT0 + Time::step( to );
+  const Time& origin = get_origin();
+
+  // We fire all spikes with time stamps up to including sliceT0 + to
+  while ( S_.position_ < P_.spike_stamps_.size() )
+  {
+    const Time tnext_stamp = origin + P_.spike_stamps_[ S_.position_ ];
+
+    // this might happen due to wrong usage of the generator
+    if ( tnext_stamp <= tstart )
+    {
+      ++S_.position_;
+      continue;
+    }
+    if ( tnext_stamp > tstop )
+    {
+      break;
+    }
+
+    long step = tnext_stamp.get_steps();
+
+    if ( get_t_min_() < step and step <= get_t_max_() )
+    {
+      SpikeEvent* se;
+      se = new SpikeEvent;
+
+      if ( P_.precise_times_ )
+      {
+        se->set_offset( P_.spike_offsets_[ S_.position_ ] );
+      }
+
+      if ( not P_.spike_multiplicities_.empty() )
+      {
+        se->set_multiplicity( P_.spike_multiplicities_[ S_.position_ ] );
+      }
+
+      // we need to subtract one from stamp which is added again in send()
+      long lag = Time( tnext_stamp - sliceT0 ).get_steps() - 1;
+
+      // all spikes are sent locally, so offset information is always preserved
+      kernel().event_delivery_manager.send( *this, *se, lag );
+      delete se;
+    }
+
+    ++S_.position_;
+  }
+}
+
+void
```

Comment:
I don't think this is necessary any more. Since the injector behaves like a normal neuron on its output side, I cannot see why we should need this restriction. @jougs, can you remember why we introduced it?

Reply:
I don't see a reason for this restriction.